src/lib.rs

```diff
@@ -680,26 +787,44 @@ fn encode_partition(fi: &FrameInvariants, fs: &mut FrameState, cw: &mut ContextW
 
     match partition {
         PartitionType::PARTITION_NONE => {
-            // TODO(anyone): Until we have RDO-based block size decision,
-            // call rdo_mode_decision() for a partition.
-            let rdo_none = rdo_mode_decision(fi, fs, cw, bsize, bo);
-            cw.bc.set_mode(bo, bsize, rdo_none.pred_mode);
+            let pred_mode = if rdo_output.part_modes.len() > 0 { 
+                rdo_output.part_modes[0].pred_mode
+            } else { // edges
+                rdo_mode_decision(fi, fs, cw, bsize, bo).part_modes[0].pred_mode };
 
-            encode_block(fi, fs, cw, rdo_none.pred_mode, bsize, bo);
+            cw.bc.set_mode(bo, bsize, pred_mode);
+
+            // FIXME every final block is encoded twice, once for RDO decision and once here
+            encode_block(fi, fs, cw, pred_mode, bsize, bo);
         },
         PartitionType::PARTITION_SPLIT => {
-            assert!(subsize != BlockSize::BLOCK_INVALID);
-            encode_partition(fi, fs, cw, subsize, bo);
-            encode_partition(fi, fs, cw, subsize, &BlockOffset{x: bo.x + hbs as usize, y: bo.y});
-            encode_partition(fi, fs, cw, subsize, &BlockOffset{x: bo.x, y: bo.y + hbs as usize});
-            encode_partition(fi, fs, cw, subsize, &BlockOffset{x: bo.x + hbs as usize, y: bo.y + hbs as usize});
+            if rdo_output.part_modes.len() >= 4 {
+                // Assume a partition encoded in a decision (so we know the optimal prediction modes)
+                assert!(subsize != BlockSize::BLOCK_INVALID);
+
+                for mode in rdo_output.part_modes {
+                    let offset = mode.bo.clone();
+
+                    encode_partition(fi, fs, cw, subsize, &offset, 
+                        &Some(RDOOutput { 
+                            rd_cost: mode.rd_cost, 
+                            part_type: PartitionType::PARTITION_NONE, 
+                            part_modes: vec![mode] }));
+                }
+            }
+            else {
+                encode_partition(fi, fs, cw, subsize, bo, &None);
```

Comment:
As record or comments for readers, this block of splits is done when input partition has not come from rdo_partition_decision(), i.e. 64x64 -> 32->32 or all unconditional 4x4s on right/bottom frame border.

Reply:
This should be addressed with the new comments added to this function.